drivers/stepper/adi_tmc/adi_tmc2209_stepper_controller.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Fabian Blatz <fabianblatz@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_tmc2209
+
+#include <zephyr/stepper/step_dir_stepper_controller.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(tmc2209, CONFIG_STEPPER_LOG_LEVEL);
+
+struct tmc2209_config {
+	struct step_dir_stepper_common_config common;
+	const struct gpio_dt_spec enable_pin;
+	const struct gpio_dt_spec *msx_pins;
+};
+
+struct tmc2209_data {
+	struct step_dir_stepper_common_data common;
+	enum stepper_micro_step_resolution resolution;
+};
+
+STEP_DIR_STEPPER_STRUCT_CHECK(struct tmc2209_config, struct tmc2209_data);
+
+static int tmc2209_stepper_enable(const struct device *dev, const bool enable)
+{
+	const struct tmc2209_config *config = dev->config;
+
+	LOG_DBG("Stepper motor controller %s %s", dev->name, enable ? "enabled" : "disabled");
+	if (enable) {
+		return gpio_pin_set_dt(&config->enable_pin, 1);
+	} else {
+		return gpio_pin_set_dt(&config->enable_pin, 0);
+	}
+}
+
+static int tmc2209_stepper_set_micro_step_res(const struct device *dev,
+					      enum stepper_micro_step_resolution micro_step_res)
+{
+	struct tmc2209_data *data = dev->data;
+	const struct tmc2209_config *config = dev->config;
+	uint8_t ms_pins_state[2];
+	int ret;
+
+	if (!config->msx_pins) {
+		LOG_ERR("Microstep resolution pins are not configured");
+		return -ENODEV;
+	}
+
+	switch (micro_step_res) {
+	case STEPPER_MICRO_STEP_8:
+		ms_pins_state[0] = 0;
+		ms_pins_state[1] = 0;
+		break;
+	case STEPPER_MICRO_STEP_16:
+		ms_pins_state[0] = 1;
+		ms_pins_state[1] = 1;
+		break;
+	case STEPPER_MICRO_STEP_32:
+		ms_pins_state[0] = 1;
+		ms_pins_state[1] = 0;
+		break;
+	case STEPPER_MICRO_STEP_64:
+		ms_pins_state[0] = 0;
+		ms_pins_state[1] = 1;
+		break;
+	default:
+		LOG_ERR("Unsupported microstep resolution: %d", micro_step_res);
+		return -EINVAL;
+	}
+
+	for (uint8_t i = 0; i < 2; i++) {
+		ret = gpio_pin_set_dt(&config->msx_pins[i], ms_pins_state[i]);
+		if (ret < 0) {
+			LOG_ERR("Failed to set MS%d pin", i + 1);
+			return ret;
+		}
+	}
+
+	data->resolution = micro_step_res;
+
+	return 0;
+}
+
+static int tmc2209_stepper_get_micro_step_res(const struct device *dev,
+					      enum stepper_micro_step_resolution *micro_step_res)
+{
+	struct tmc2209_data *data = dev->data;
+
+	*micro_step_res = data->resolution;
+	return 0;
+}
+
+static const struct stepper_driver_api tmc2209_stepper_api = {
```

Comment:
static struct DEVICE_API.

Also can this be moved right above TMC2209 Define?

Reply:
Done